log/log.go

```diff
@@ -23,14 +23,32 @@ func With(logger Logger, keyvals ...interface{}) Logger {
 	if w, ok := logger.(Wither); ok {
 		return w.With(keyvals...)
 	}
-	return LoggerFunc(func(kvs ...interface{}) error {
-		// Limiting the capacity of the first argument to append ensures that
-		// a new backing array is created if the slice must grow. Using the
-		// extra capacity without copying risks a data race that would violate
-		// the Logger interface contract.
-		n := len(keyvals)
-		return logger.Log(append(keyvals[:n:n], kvs...)...)
-	})
+	// Limiting the capacity of the stored keyvals ensures that a new
+	// backing array is created if the slice must grow in Log or With.
+	// Using the extra capacity without copying risks a data race that
+	// would violate the Logger interface contract.
+	n := len(keyvals)
+	return &withLogger{
+		logger:  logger,
+		keyvals: keyvals[:n:n],
+	}
+}
+
+type withLogger struct {
+	logger  Logger
+	keyvals []interface{}
+}
+
+func (l *withLogger) Log(keyvals ...interface{}) error {
+	return l.logger.Log(append(l.keyvals, keyvals...)...)
```

Comment:
Yes, `l.keyvals` is capacity sliced when each `withLogger` is created. See line 33 and 50. The code passes `TestWithConcurrent` with race detector and multiple cores enabled.